### DISCUSSION TITLE
how to declaratively round values on serialization? can it be dynamic?

### DISCUSSION THREAD

**Discussion by tamird** [pydantic/pydantic#8881] [Question]
Suppose I have a model with a `Decimal` field which I want to have e.g. `decimal_places=2`. On parsing, I want this validated - but on construction I want to implement a rounding policy. Can pydantic support this?

Further: can a field's validation depend on another field? Consider currency, naively:

```
Class Money(BaseModel)
  amount: Decimal
  currency: str
```
if currency is "USD", I'd want amount rounded to 2 decimal places. But if currency is "BTC" I might want considerably more. How can I implement this?

Note: apart from currency I have other data types with similar needs like units of distance (meters, kilometers) which want different rounding precision - so I'm searching for a solution that is largely reusable across types of quantities.

Thanks!

**Comment by H-G-11** [comment]
You should use [model validators](https://docs.pydantic.dev/latest/concepts/validators/#model-validators). I don't think you need [Pydantic's decimal](https://docs.pydantic.dev/latest/api/standard_library_types/#decimaldecimal), float should be fine if I understand your use-case.

But for validation you will need Python Decimal package, see the first examples [here](https://docs.python.org/3/library/decimal.html#decimal-faq).

Example
```python
from enum import StrEnum
from typing import Any
from decimal import Decimal, Context, Inexact

from pydantic import BaseModel, model_validator

class Currency(StrEnum):
    usd = "usd"
    btc = "btc"

CURRENCY_TO_ROUNDING: dict[Currency, int] = {
    Currency.usd: 2,
    Currency.btc: 20,
}

class Money(BaseModel):
    amount: float
    currency: str

    @model_validator(mode='before')
    @classmethod
    def validate_and_round_amount(cls, data: Any) -> Any:
        # you can do additional validation on the dictionary if you want
        ...
        rounding_decimals = CURRENCY_TO_ROUNDING.get(data["currency"], 2)
        PLACES = Decimal(10) ** -rounding_decimals
        # will raise error if needed
        Decimal(data["amount"]).quantize(PLACES, context=Context(traps=[Inexact]))
        data["amount"] = round(Decimal(data["amount"]), rounding_decimals)
        return data
```

This works
```python
Money(amount="2.31", currency="usd")
Money(amount="2.312323", currency="btc")
```

This does not
```python
Money(amount="2.312323", currency="usd")
```

The validator is flexible enough so that you can handle all kind of scenarios, like different rounding strategies etc.

**Reply by tamird** [reply]
> Note: apart from currency I have other data types with similar needs like units of distance (meters, kilometers) which want different rounding precision - so I'm searching for a solution that is largely reusable across types of quantities.

This doesn't address this need.

**Comment by Viicos** [comment]
> Further: can a field's validation depend on another field?

Yes, as noted by @H-G-11, using model validators (I'd recommend using the `mode='after'` however).

> On parsing, I want this validated - but on construction I want to implement a rounding policy.

Could you clarify _on parsing_ vs _on construction_?

**Reply by tamird** [reply]
Construction would be directly building the object `MyClass(field_a=a)` and parsing would be using `MyClass.model_validate_json`.
